contrib/gazebo/gazebo_sawyer_env.py

```diff
@@ -0,0 +1,419 @@
+    return np.linalg.norm(goal_a - goal_b, axis=-1)
+
+def rate_limited(max_per_second: int):
+    """Rate-limits the decorated function locally, for one process."""
+    lock = threading.Lock()
+    min_interval = 1.0 / max_per_second
+
+    def decorate(func):
+        last_time_called = time.perf_counter()
+
+        @wraps(func)
+        def rate_limited_function(*args, **kwargs):
+            lock.acquire()
+            nonlocal last_time_called
+            try:
+                elapsed = time.perf_counter() - last_time_called
+                left_to_wait = min_interval - elapsed
+                if left_to_wait > 0:
+                    time.sleep(left_to_wait)
+
+                return func(*args, **kwargs)
+            finally:
+                last_time_called = time.perf_counter()
+                lock.release()
+
+        return rate_limited_function
+
+    return decorate
+
+class GazeboSawyerEnv(GazeboEnv, Serializable):
+    """
+    Gazebo Sawyer Env
+    """
+
+    def __init__(self, has_object, distance_threshold,
+                 initial_qpos, reward_type,  target_in_the_air,
+                 limb='right', tip_name="right_gripper_tip", target_range=0.15,
+                 obj_range=0.15):
+        """
+        Input
+        -----
+        control_mode: have 4 different control modes: position, velocity, trajectory-position, torque
+        """
+        # run the superclass initialize function first
+        Serializable.quick_init(self, locals())
+
+        # launch gazebo environment
+        # launchpath = osp.join(rospkg.RosPack().get_path('sawyer_gazebo'), "launch")
+        # launch_file = osp.join(launchpath, 'sawyer_learning.launch')
+        # if not osp.exists(launch_file):
+        #     raise IOError("File %s" % launch_file)
+        # subprocess.Popen(["roslaunch", launch_file])
+        # rospy.loginfo("Gazebo launched!")
+        rospy.init_node("GazeboSawyerPosControlEnv", anonymous=True)
+        # Verify robot is enabled
+        print("Getting robot state...")
+        self._rs = intera_interface.RobotEnable(intera_interface.CHECK_VERSION)
+        self._init_state = self._rs.state().enabled
+        print("Enabling robot...")
+        self._rs.enable()
+
+        def clean_shutdown():
+            # rospy.wait_for_service('/gazebo/unpause_physics')
+            # try:
+            #     self.unpause()
+            # except rospy.ServiceException as e:
+            #     print('/gazebo/unpause_physics service call failed')
+
+            print("Exiting...")
+            if self._rs.state().enabled:
+                print("Disabling robot...")
+                self._rs.disable()
+            # delete model
+            print("Delete gazebo models...")
+            GazeboEnv.delete_gazebo_model(model_name="cafe_table")
+            GazeboEnv.delete_gazebo_model(model_name="block")
+            GazeboEnv.delete_gazebo_model(model_name="target")
+
+            # Kill gzclient, gzserver and roscore
```

Comment:
Is there a case where we need the following deactivated code? If so, you should explain when to use it or else remove it.